inc/class-s3-uploads.php

```diff
@@ -277,4 +352,168 @@ public function copy_image_from_s3( $file ) {
+				'Key' => $location['key'],
+				'ACL' => $acl,
+			] );
+		}
+
+		try {
+			Aws\CommandPool::batch( $s3, $commands );
+		} catch ( Exception $e ) {
+			return new WP_Error( $e->getCode(), $e->getMessage() );
+		}
+
+		return null;
+	}
+
+	/**
+	 * Get all the files stored for a given attachment.
+	 *
+	 * @param integer $attachment_id
+	 * @return array Array of all full paths to the attachment's files.
+	 */
+	public static function get_attachment_files( int $attachment_id ) : array {
+		$uploadpath = wp_get_upload_dir();
+		$main_file = get_attached_file( $attachment_id );
+		$files = [ $main_file ];
+		$meta = wp_get_attachment_metadata( $attachment_id );
+		if ( isset( $meta['sizes'] ) ) {
+			foreach ( $meta['sizes'] as $size => $sizeinfo ) {
+				$files[] = $uploadpath['basedir'] . $sizeinfo['file'];
+			}
+		}
+
+		$original_image = get_post_meta( $attachment_id, 'original_image', true );
+		if ( $original_image ) {
+			$files[] = $uploadpath['basedir'] . $original_image;
+		}
+
+		$backup_sizes = get_post_meta( $attachment_id, '_wp_attachment_backup_sizes', true );
+		if ( $backup_sizes ) {
+			foreach ( $backup_sizes as $size => $sizeinfo ) {
+				// Backup sizes only store the backup filename, which is relative to the
+				// main attached file, unlike the metadata sizes array.
+				$files[] = path_join( dirname( $main_file ), $sizeinfo['file'] );
+			}
+		}
+
+		$files = apply_filters( 's3_uploads_get_attachment_files', $files, $attachment_id );
+
+		return $files;
+	}
+
+	/**
+	 * Add the S3 signed params onto an image for for a given attachment.
+	 *
+	 * This function determines whether the attachment needs a signed URL, so is safe to
+	 * pass any URL.
+	 *
+	 * @param string $url
+	 * @param integer $post_id
+	 * @return string
+	 */
+	public function add_s3_signed_params_to_attachment_url( string $url, int $post_id ) : string {
+		if ( ! $this->is_private_attachment( $post_id ) ) {
+			return $url;
+		}
+		$path = $this->get_s3_location_for_url( $url );
+		if ( ! $path ) {
+			return $url;
+		}
+		$cmd = $this->s3()->getCommand('GetObject', [
+			'Bucket' => $path['bucket'],
+			'Key' => $path['key'],
+		]);
+
+		$presigned_url_expires = apply_filters( 's3_uploads_private_attachment_url_expiry', '+6 hours', $post_id );
+		$query = $this->s3()->createPresignedRequest( $cmd, $presigned_url_expires )->getUri()->getQuery();
+
+		// The URL could have query params on it already (such as being an already signed URL),
+		// but query params will mean the S3 signed URL will become corrupt. So, we have to
+		// remove all query params.
+		$url = strtok( $url, '?' ) . '?' . $query;
+		$url = apply_filters( 's3_uploads_presigned_url', $url, $post_id );
+
+		return $url;
+	}
+
+	/**
+	 * Add the S3 signed params to an image src array.
+	 *
+	 * @param array|false $image
+	 * @param integer $post_id
+	 * @return array|false
+	 */
+	public function add_s3_signed_params_to_attachment_image_src( $image, int $post_id ) {
```

Comment:
@joehoyle So we didn't take into the account that `wp_get_attachment_image_src` can accept both `array|false` as seen here https://developer.wordpress.org/reference/functions/wp_get_attachment_image_src/ I actually came across this on the project.

Reply:
Good catch, thanks @stuartshields